include/core/non_linear_solver.h

```diff
@@ -46,7 +43,9 @@ class NonLinearSolver
                   const Parameters::NonLinearSolver &params);
 
   virtual ~NonLinearSolver()
-  {}
+  { // Initialized the outer_iteration_counter by default to 0.
+    outer_iteration = 0;
+  }
```

Comment:
This should be put in the  pre-constructor and not the constructor
After the : statement and not within {}